### ISSUE TITLE
False nullability warnings caused by discard patterns that don't change the code flow

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#51800]
**Version Used**: 
```
Branch main (8 Mar 2021)
Latest commit eddde7c by Sam Harwell:
Merge pull request #51559 from sharwell/current-thread

Use Environment.CurrentManagedThreadId
```

**Steps to Reproduce**:

```cs
#nullable enable

public class Test
{
  public bool Unrelated => false;
  public string? Value { get; set; }

  public string M1()
  {
    switch (this)
    {
      case { Value: _, Unrelated: _ } x:
        return "";
      default:
        return Value; // no warnings
    }
  }
    
  public string M2()
  {
    switch (this)
    {
      case { Unrelated: _ } x:
        return "";
      default:
        return Value; // False CS8603: Possible null reference return.
    }
  }
}
```

**Expected Behavior**:
No warnings in both examples. The `default` sections of both switch statements aren't reachable

**Actual Behavior**:
False `CS8603: Possible null reference return` is reported for `M2` but not `M1`

**Comment by Youssef1313** [comment]
An unrealistic edge case to consider where a similar code can reach the `default` case:

```
#nullable enable

using System;

var s = new Test().M2();
Console.WriteLine(s is null);
Console.WriteLine(s.ToString());

public class Test
{
  public int Length => 0;
  public string? Value { get; set; }

  public static implicit operator string?(Test c) => null;
    
  public string M2()
  {
    switch (this)
    {
      case { Length: _ } x:
        return "";
      default:
        return Value;
    }
  }
}
```

[SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8DEMqANsRMMQKYAElM5VAsAFAsACATAIwssBuEZNQDO1ALzUYlAO7UAKpWEIAFAEoAdAFkOagNzsuATmWioooqVX7mbIyfVyA9gGUEyWAHM1V3jYDM1JzyiggsAN4s1IEBsAjUADJ0HggAFuIAfNQADNZRbAG2WQD81ABqEMSoNGHUHpQIuiL1jQC+vnkFXABs1FBgAA7EUADGUHGO/ZTIEAiOQoVFygpK1MOqGZIkxLlR1JHRgVxZ1Npq+xHMuyLSY8NpyqlmqvtRF1dRwxDC1QlJqSDUAD61Ba1AAHiAXu82AB2agAInhOyuGEoADMICQkFCrrCyhUqsiQfs2sxSUA).

**Comment by RikkiGibson** [comment]
I think I would expect a warning in both examples, because the nullability of `Value` is never tested in either of them. It concerns me that its state apparently changed to not-null at the point where it was returned in one of the examples.
